fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -0,0 +1,519 @@
+                                protoEx.initCause(ex);
+                                throw protoEx;
+                            }
+                            if ( savedStamp == null || ! matchingIndexingTypeStamp(savedStamp)) {
+                                throw new RecordCoreException("This index was partly built by another method",
+                                        LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
+                                        LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
+                                        LogMessageKeys.INDEXER_ID, common.getUuid(),
+                                        LogMessageKeys.EXPECTED, indexingTypeStamp,
+                                        LogMessageKeys.ACTUAL, savedStamp);
+                            }
+                        }
+                        return null;
+                    });
+        } ));
+    }
+
+    @Nonnull
+    abstract IndexBuildProto.IndexBuildIndexingStamp getIndexingTypeStamp();
+
+    @Nonnull
+    private boolean matchingIndexingTypeStamp(@Nonnull final IndexBuildProto.IndexBuildIndexingStamp stamp) {
+        return getIndexingTypeStamp().equals(stamp);
+    }
+
+    abstract CompletableFuture<Void> buildIndexInternalAsync();
+
+    // Helpers for implementing modules. Some of them are public to support unit-testing.
+    protected void maybeLogBuildProgress(SubspaceProvider subspaceProvider, List<Object> additionalLogMessageKeyValues) {
+        if (LOGGER.isInfoEnabled() && shouldLogBuildProgress()) {
+            final Index index = common.getIndex();
+            LOGGER.info(KeyValueLogMessage.build("Built Range",
+                    LogMessageKeys.INDEX_NAME, index.getName(),
+                    LogMessageKeys.INDEX_VERSION, index.getLastModifiedVersion(),
+                    subspaceProvider.logKey(), subspaceProvider,
+                    LogMessageKeys.RECORDS_SCANNED, common.getTotalRecordsScanned().get(),
+                    LogMessageKeys.INDEXER_ID, common.getUuid())
+                    .addKeysAndValues(additionalLogMessageKeyValues)
+                    .toString());
+        }
+    }
+
+    private boolean shouldLogBuildProgress() {
+        long interval = common.config.getProgressLogIntervalMillis();
+        long now = System.currentTimeMillis();
+        if (interval == 0 || interval < (now - timeOfLastProgressLogMillis)) {
+            return false;
+        }
+        timeOfLastProgressLogMillis = now;
+        return true;
+    }
+
+    public int getLimit() {
+        // made public to support tests
+        return throttle.getLimit();
+    }
+
+    protected CompletableFuture<Boolean> throttleDelay() {
+        int limit = getLimit();
+        int recordsPerSecond = common.config.getRecordsPerSecond();
+        int toWait = (recordsPerSecond == IndexingCommon.UNLIMITED) ? 0 : 1000 * limit / recordsPerSecond;
```

Comment:
Jitter?

Reply:
This delay is there to try and control the rate at which records are being indexed to avoid overwhelming the cluster with writes; it is not a delay added as part of an internal retry. The delay added when there's a retry does have jitter (see: https://github.com/FoundationDB/fdb-record-layer/blob/249a83a0b1664234290da3142292c54401803eaf/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingThrottle.java#L222), though it could probably be better.

Reply:
This part is throttling for scanning speed, so I don't think jittering will be necessary. We are doing jitter when we back off on failures, see https://github.com/FoundationDB/fdb-record-layer/blob/main/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java#L433

Reply:
Actually, if you are wondering about because of the backoff problem I mentioned this morning, it's at https://github.com/FoundationDB/fdb-record-layer/issues/1007

Reply:
Sorry, I'm not sure I understand the question. Can you clarify?